lib/vanagon/engine/hardware.rb

```diff
@@ -0,0 +1,102 @@
+require 'vanagon/engine/base'
+require 'json'
+require 'redis'
+
+REDIS_HOST = 'redis.delivery.puppetlabs.net'
+REDIS_PORT = 6379
+LOCK_USER = "vanagon/#{ENV['USER']}"
+
+
+class Vanagon
+  class Engine
+    # Class to use when building on a hardware device (e.g. AIX, Switch, etc)
+    #
+    class Hardware < Base
+
+      # This method is used to obtain a vm to build upon
+      # For the base class we just return the target that was passed in
+      def select_target
+        @build_host
+      end
+
+      # Steps needed to tear down or clean up the system after the build is
+      # complete. In this case, we'll attempt to unlock the hardware
+      def teardown
+        release_lock
+      end
+
+      def initialize(platform)
+        # see if we can aquire a lock
+        Vanagon::Driver.logger.debug "Hardware engine invoked."
+        @platform = platform
+        super
+        @required_attributes << "build_host"
+        @build_host = platform.build_host
+        @target = @build_host
+        aquire_lock
+      end
+
+      # Lock the build_host as specificied in the platform definition. This
+      # implement a rudimentary backoff sequence in order to wait on the host
+      # to unlock. Note: this polls forever.
+      def aquire_lock
+        # Wait if locked
+        if locked?
+          warn "Unable to aquire lock."
+        end
+        sleepcount = 1
+        psleepcount = 1
+        temp = 0
+        while locked?
+          print "#{@build_host} is locked..."
+          puts "waiting #{sleepcount} seconds."
+          sleep sleepcount
+          temp = sleepcount
+          sleepcount += psleepcount
+          psleepcount = temp
+        end
+        lock_contents = {}
+        lock_contents[:user] = LOCK_USER
+        lock_contents[:time] = Time.now.to_s
+        lock_contents[:reason] = "Vanagon build on #{@build_host}"
+        redis = connect_redis
+        redis.set "node_lock_#{@build_host.split('.')[0]}", lock_contents.to_json
+        puts "Hardware locked: #{@build_host}"
+        Vanagon::Driver.logger.info "Hardware locked: #{@build_host}"
+      end
+
+      # Release the lock on build_host
+      def release_lock
+        redis = connect_redis
+        r = redis.get("node_lock_#{@build_host.split('.')[0]}")
```

Comment:
Yeah, this was an issue I ran into when testing. I think all locks should split at the dot. Otherwise a lock on pe-aix-71-builder won't match pe-aix-71-builder.delivery.puppetlabs.net. That's not so good. I was going to jump in the kerminator code and fix that.